File: 向量检索原理/度量方法算法实现/cosine_similarity.cpp
// cosine_similarity.cpp
#include <vector>
#include <cmath>
#include <iostream>

double dot_product(const std::vector<double>& vec1, const std::vector<double>& vec2) {
    double dot = 0.0;
    for (size_t i = 0; i < vec1.size(); ++i) {
        dot += vec1[i] * vec2[i];
    }
    return dot;
}

double magnitude(const std::vector<double>& vec) {
    double sum = 0.0;
    for (double val : vec) {
        sum += val * val;
    }
    return std::sqrt(sum);
}

double cosine_similarity(const std::vector<double>& vec1, const std::vector<double>& vec2) {
    return dot_product(vec1, vec2) / (magnitude(vec1) * magnitude(vec2));
}

int main() {
    std::vector<double> vec1 = {1.0, 2.0, 3.0};
    std::vector<double> vec2 = {4.0, 5.0, 6.0};
    std::cout << "Cosine Similarity: " << cosine_similarity(vec1, vec2) << std::endl;
    return 0;
}

File: 向量检索原理/度量方法算法实现/cosine_similarity.py
# cosine_similarity.py
import math

def dot_product(vec1, vec2):
    return sum(a * b for a, b in zip(vec1, vec2))

def magnitude(vec):
    return math.sqrt(sum(x ** 2 for x in vec))

def cosine_similarity(vec1, vec2):
    return dot_product(vec1, vec2) / (magnitude(vec1) * magnitude(vec2))

if __name__ == "__main__":
    vec1 = [1.0, 2.0, 3.0]
    vec2 = [4.0, 5.0, 6.0]
    print("Cosine Similarity:", cosine_similarity(vec1, vec2))

